### PR TITLE
5746 fixing reducer step

### LINKED ISSUE #5746: completion handler not fired for jobs with reducer step
When a job has a reducer step, the completion handler is never fired

There's also an issue with consume step and the singleton nature of reduction jobs that needs to be addressed (specifically in bulk export)

### PR DESCRIPTION
**Pull request by TipzCM**
closes #5746

### CODE REVIEW

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/coordinator/ReductionStepExecutorServiceImpl.java [hunk 1]
```diff
@@ -256,6 +257,18 @@ ReductionStepChunkProcessingResponse executeReductionStep(
...
 				}
+
+				if (response.isSuccessful()) {
+					/**
+					 * All reduction steps are final steps.
+					 */
+					IJobCompletionHandler<PT> completionHandler =
+							theJobWorkCursor.getJobDefinition().getCompletionHandler();
+					if (completionHandler != null) {
+						completionHandler.jobComplete(new JobCompletionDetails<>(parameters, instance));
```
Comment (codeforgreen) [inline hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/coordinator/ReductionStepExecutorServiceImpl.java:268]: This looks fine to me. But I am just curious why `JobInstanceStatusUpdater` `updateInstanceStatus` with COMPLETED is not called.

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/batch2/Batch2CoordinatorIT.java [hunk 2]
```diff
@@ -231,76 +233,122 @@ public void testFirstStepToSecondStep_singleChunkFasttracks() throws Interrupted
...
+				SecondStepOutput output = new SecondStepOutput();
+				output.setValue(testInfo + secondStepInt.getAndIncrement());
+				theDataSink.accept(output);
+			}
 
+			@Override
+			public void reductionStepConsume(ChunkExecutionDetails<TestJobParameters, SecondStepOutput> theChunkDetails, IJobDataSink<ReductionStepOutput> theDataSink) {
+				int val = mySecondGate.getAndIncrement();
+				if (val == (totalCalls - 1)) {
+					myBatch2JobHelper.forceRunMaintenancePass();
```
Comment (codeforgreen) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/batch2/Batch2CoordinatorIT.java:275]: Why is this necessary?

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/batch2/Batch2CoordinatorIT.java [hunk 3]
```diff
@@ -551,6 +592,80 @@ private JobDefinition<? extends IModelJson> buildGatedJobDefinition(String theJo
...
+			.setParametersType(TestJobParameters.class)
+			.gatedExecution()
+			.addFirstStep(
+				FIRST_STEP_ID,
+				"Test first step",
+				FirstStepOutput.class,
+				theFirstStep
+			)
+			.addIntermediateStep("SECOND",
+				"Second step",
```
Comment (codeforgreen) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/batch2/Batch2CoordinatorIT.java:655]: Minor: to be consistent, create a constant for this, as the first and the last steps have that already.

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/batch2/Batch2CoordinatorIT.java [hunk 4]
```diff
@@ -551,6 +592,80 @@ private JobDefinition<? extends IModelJson> buildGatedJobDefinition(String theJo
...
+			.setJobDefinitionVersion(TEST_JOB_VERSION)
+			.setParametersType(TestJobParameters.class)
+			.gatedExecution()
+			.addFirstStep(
+				FIRST_STEP_ID,
+				"Test first step",
+				FirstStepOutput.class,
+				theFirstStep
+			)
+			.addIntermediateStep("SECOND",
```
Comment (codeforgreen) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/batch2/Batch2CoordinatorIT.java:654]: Minor: to be consistent, create a constant for this, as the first and the last steps have that already.

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/batch2/Batch2CoordinatorIT.java (hunk 2 above)
Comment (TipzCM) [reply hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/batch2/Batch2CoordinatorIT.java:275]: It was copied over from a time when the reduction step worked differently.

I guess it isn't needed anymore (i will remove it)

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/batch2/Batch2CoordinatorIT.java (hunk 4 above)
Comment (TipzCM) [reply hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/batch2/Batch2CoordinatorIT.java:654]: updated

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/7_2_0/5746-reduction-jobs-should-fire-completion-handler.yaml [hunk 5]
```diff
@@ -0,0 +1,6 @@
+---
+type: fix
+issue: 5746
+title: "Batch2 jobs with reduction steps didn't fire the completion handler.
```
Comment (lukedegruchy) [inline hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/7_2_0/5746-reduction-jobs-should-fire-completion-handler.yaml:4]: Please a closes link in your PR

**Review by lukedegruchy — APPROVED** [review]

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/7_2_0/5746-reduction-jobs-should-fire-completion-handler.yaml (hunk 5 above)
Comment (TipzCM) [reply hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/7_2_0/5746-reduction-jobs-should-fire-completion-handler.yaml:4]: added

**Pull request merged**